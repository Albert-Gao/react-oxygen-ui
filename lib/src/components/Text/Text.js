"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const react_1 = __importDefault(require("react"));
const styled_system_1 = require("styled-system");
const fontSpecs_1 = require("../../styles/theme/constants/fontSpecs");
const utils_1 = require("../../utils/utils");
const BaseBox_1 = require("../Box/BaseBox");
function addStyleToResult(tempStyle, result, key) {
    if (utils_1.is.notExist(result[key])) {
        result[key] = [];
    }
    // @ts-ignore
    // tslint:disable-next-line: no-unsafe-any
    result[key].push(tempStyle[key]);
}
exports.getFontSpecStyle = (noDefaultLineHeight, isMaxWidthIncluded = false, size) => {
    if (utils_1.is.notExist(size)) {
        return {};
    }
    let result = {};
    if (Array.isArray(size)) {
        if (size.length === 1) {
            result = exports.getFontSpecStyle(noDefaultLineHeight, isMaxWidthIncluded, size[0]);
        }
        else {
            const temp = [];
            size.forEach(s => {
                const style = exports.getFontSpecStyle(noDefaultLineHeight, isMaxWidthIncluded, s);
                temp.push(style);
            });
            temp.forEach(tempStyle => {
                addStyleToResult(tempStyle, result, 'fontSize');
                if (!noDefaultLineHeight) {
                    addStyleToResult(tempStyle, result, 'lineHeight');
                }
                if (isMaxWidthIncluded) {
                    addStyleToResult(tempStyle, result, 'maxWidth');
                }
            });
        }
    }
    else {
        const fontSpecInTheme = fontSpecs_1.fontSpecs[size];
        const isInTheme = typeof fontSpecInTheme === 'object';
        if (isInTheme) {
            result = {
                fontSize: styled_system_1.px(fontSpecInTheme.size),
                lineHeight: fontSpecInTheme.lineHeight,
            };
            if (isMaxWidthIncluded) {
                result.maxWidth = styled_system_1.px(fontSpecInTheme.maxWidth);
            }
        }
        else {
            result = {
                fontSize: styled_system_1.px(size),
            };
        }
    }
    if (result === {}) {
        result = { fontSize: size };
    }
    return result;
};
exports.Text = ({ size, noDefaultLineHeight = false, takeRef, ...rest }) => (<BaseBox_1.BaseBox as="div" {...exports.getFontSpecStyle(noDefaultLineHeight, false, size)} {...utils_1.getRefProp(takeRef)} {...rest}/>);
