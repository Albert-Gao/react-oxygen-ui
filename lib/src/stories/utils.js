"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const addon_actions_1 = require("@storybook/addon-actions");
const addon_knobs_1 = require("@storybook/addon-knobs");
const styles_1 = require("../styles");
function arrayToObject(data, keyToMap) {
    if (!data || data.length === 0) {
        return {};
    }
    const result = {};
    data.forEach((item, index) => {
        const i = keyToMap ? keyToMap[index] : index;
        result[i] = item;
    });
    return result;
}
exports.arrayToObject = arrayToObject;
// We wrap the action to prevent been display in addon-jsx
function wrapAction(actionName) {
    const wrapped = addon_actions_1.action(actionName);
    wrapped.toString = () => `action('${actionName}')`;
    return wrapped;
}
exports.wrapAction = wrapAction;
const keyOfColors = Object.keys(styles_1.defaultTheme.colors);
exports.Knobs = {
    spacesKnob: (propName, defaultValue = 2) => addon_knobs_1.select(propName, styles_1.DEFAULT_SPACE, defaultValue),
    colorKnob: (propName, defaultValue = 'primary') => addon_knobs_1.select(propName, arrayToObject(keyOfColors, keyOfColors), defaultValue),
    fontWeightKnob: (defaultValue = 'regular') => addon_knobs_1.select('fontWeight', {
        thin: 'thin',
        light: 'light',
        regular: 'regular',
        medium: 'medium',
        bold: 'bold',
        black: 'black',
    }, defaultValue),
    fontSizeKnob: (propName = 'fontSize', defaultValue = 1) => addon_knobs_1.select(propName, styles_1.DEFAULT_FONT_SIZE, defaultValue),
    sizeKnob: (propName, defaultValue = 'medium') => addon_knobs_1.select(propName, styles_1.DEFAULT_SIZE, styles_1.DEFAULT_SIZE[defaultValue]),
};
