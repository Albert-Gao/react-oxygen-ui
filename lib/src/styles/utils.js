"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const styled_system_1 = require("styled-system");
const utils_1 = require("../utils/utils");
exports.getMediaQuery = (screenSize, css) => `${styled_system_1.createMediaQuery(screenSize)} {
    ${css};
  }`;
exports.getTransitionValue = (propertyNames = 'transform') => `${propertyNames} 0.3s cubic-bezier(0.53, 0.01, 0.36, 1.63);`;
function getSafeValue(target, defaultValue) {
    return utils_1.is.notExist(target) ? defaultValue : target;
}
exports.getSafeValue = getSafeValue;
function getSafeCss(target) {
    return getSafeValue(target, '');
}
exports.getSafeCss = getSafeCss;
