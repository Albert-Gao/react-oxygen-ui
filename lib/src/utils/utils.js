"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.is = {
    function: (v) => typeof v === 'function',
    number: (v) => typeof v === 'number',
    notExist: (v) => v === undefined || v === null,
};
exports.getRefProp = (takeRef) => takeRef
    ? {
        ref: takeRef,
    }
    : {};
