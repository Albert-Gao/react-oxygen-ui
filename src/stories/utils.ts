import { action } from '@storybook/addon-actions';
import { select } from '@storybook/addon-knobs';
import { Responsive } from '../components';
import {
  ColorType,
  DEFAULT_FONT_SIZE,
  DEFAULT_SIZE,
  DEFAULT_SPACE,
  defaultTheme,
  FontWeightType,
} from '../styles';

export function arrayToObject<T>(
  data: T[],
  keyToMap?: string[],
): { [id: string]: T } {
  if (!data || data.length === 0) {
    return {};
  }

  const result: { [id: string]: T } = {};
  data.forEach((item, index) => {
    const i = keyToMap ? keyToMap[index] : index;
    result[i] = item;
  });

  return result;
}

// We wrap the action to prevent been display in addon-jsx
export function wrapAction(actionName: string) {
  const wrapped = action(actionName);
  wrapped.toString = () => `action('${actionName}')`;

  return wrapped;
}

const keyOfColors = Object.keys(defaultTheme.colors) as ColorType[];

export const Knobs = {
  spacesKnob: (propName: string, defaultValue: number = 2) =>
    select<number>(propName, DEFAULT_SPACE, defaultValue),

  colorKnob: (
    propName: string,
    defaultValue: ColorType = 'primary',
  ) =>
    select<string>(
      propName,
      arrayToObject<string>(
        keyOfColors as string[],
        keyOfColors as string[],
      ),
      defaultValue,
    ),

  fontWeightKnob: (defaultValue: FontWeightType = 'regular') =>
    select(
      'fontWeight',
      {
        thin: 'thin',
        light: 'light',
        regular: 'regular',
        medium: 'medium',
        bold: 'bold',
        black: 'black',
      },
      defaultValue,
    ),

  fontSizeKnob: (
    propName: string = 'fontSize',
    defaultValue: Responsive = 1,
    // @ts-ignore
  ) => select(propName, DEFAULT_FONT_SIZE, defaultValue),

  sizeKnob: (
    propName: string,
    defaultValue: keyof typeof DEFAULT_SIZE = 'medium',
  ) => select(propName, DEFAULT_SIZE, DEFAULT_SIZE[defaultValue]),
};
