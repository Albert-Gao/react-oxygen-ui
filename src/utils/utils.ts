import { IBaseBox } from '../components';
type NullOrUndefined = null | undefined;

export const is = {
  function: (v: unknown): v is Function => typeof v === 'function',
  number: (v: unknown): v is number => typeof v === 'number',
  notExist: (v: unknown): v is NullOrUndefined =>
    v === undefined || v === null,
};

export const getRefProp = (takeRef?: IBaseBox['ref']) =>
  takeRef
    ? {
        ref: takeRef,
      }
    : {};
