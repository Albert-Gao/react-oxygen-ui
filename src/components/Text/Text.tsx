import * as React from "react";
import { getRefProp } from "../../utils/utils";
import { BaseBox } from "../Box/BaseBox";
import { IBaseBoxWithTakeRef } from "../Box/BaseBox.type";

export const Text: React.FC<IBaseBoxWithTakeRef> = ({ takeRef, ...rest }) => (
  <BaseBox as="div" {...getRefProp(takeRef)} {...rest} />
);
