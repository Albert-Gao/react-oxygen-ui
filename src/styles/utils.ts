import { FlattenInterpolation } from 'styled-components';
import { createMediaQuery } from 'styled-system';
import { CSSType } from '../components/Box/BaseBox.type';
import { is } from '../utils/utils';
import { ITheme } from './IThemeType';

export const getMediaQuery = (
  screenSize: number | string,
  css: string | FlattenInterpolation<ITheme>,
) =>
  `${createMediaQuery(screenSize)} {
    ${css};
  }`;

export const getTransitionValue = (propertyNames: string = 'transform') =>
  `${propertyNames} 0.3s cubic-bezier(0.53, 0.01, 0.36, 1.63);`;

export function getSafeValue<T>(target: unknown, defaultValue: T) {
  return is.notExist(target) ? defaultValue : target;
}

export function getSafeCss(target?: CSSType) {
  return getSafeValue(target, '') as CSSType;
}
