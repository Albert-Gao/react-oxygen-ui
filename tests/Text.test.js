import 'jest-styled-components';
import 'jest-dom/extend-expect';
import React from 'react';
import { render, cleanup } from 'react-testing-library';
import { Text } from '../src/components/Text/Text';
import { TestThemeProvider as ThemeProvider } from './utils';
import { css as cssFunc } from 'styled-components';

afterEach(cleanup);

describe('Test <Text css={}/>: ', () => {
  test('css prop should be added', () => {
    const { firstChild } = render(
      <ThemeProvider>
        <Text css={'background-color: black;'} />
      </ThemeProvider>,
    ).container;
    expect(firstChild).toMatchSnapshot();
    expect(firstChild).toHaveStyleRule('background-color', 'black');
  });

  test('css prop (object) should be handled', () => {
    const { firstChild } = render(
      <ThemeProvider>
        <Text css={{ backgroundColor: 'black', paddingLeft: '1023px' }} />
      </ThemeProvider>,
    ).container;
    expect(firstChild).toMatchSnapshot();
    expect(firstChild).toHaveStyleRule('background-color', 'black');
    expect(firstChild).toHaveStyleRule('padding-left', '1023px');
  });

  test('css prop (css string) should be handled', () => {
    const { firstChild } = render(
      <ThemeProvider>
        <Text
          css={cssFunc`
          background-color: black;
          padding-left: 1023px;
          `}
        />
      </ThemeProvider>,
    ).container;
    expect(firstChild).toMatchSnapshot();
    expect(firstChild).toHaveStyleRule('background-color', 'black');
    expect(firstChild).toHaveStyleRule('padding-left', '1023px');
  });
});
