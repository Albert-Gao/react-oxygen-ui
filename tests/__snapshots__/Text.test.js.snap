// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test <Text css={}/>:  css prop (css string) should be handled 1`] = `
.c0 {
  box-sizing: border-box;
  -webkit-transition: all 0.3s cubic-bezier(0.53,0.01,0.36,1.63);
  transition: all 0.3s cubic-bezier(0.53,0.01,0.36,1.63);
  background-color: black;
  padding-left: 1023px;
}

<div
  class="c0"
/>
`;

exports[`Test <Text css={}/>:  css prop (object) should be handled 1`] = `
.c0 {
  box-sizing: border-box;
  -webkit-transition: all 0.3s cubic-bezier(0.53,0.01,0.36,1.63);
  transition: all 0.3s cubic-bezier(0.53,0.01,0.36,1.63);
  background-color: black;
  padding-left: 1023px;
}

<div
  class="c0"
/>
`;

exports[`Test <Text css={}/>:  css prop should be added 1`] = `
.c0 {
  box-sizing: border-box;
  -webkit-transition: all 0.3s cubic-bezier(0.53,0.01,0.36,1.63);
  transition: all 0.3s cubic-bezier(0.53,0.01,0.36,1.63);
  background-color: black;
}

<div
  class="c0"
/>
`;
